### PR TITLE
[AMDGPU][True16][MC] Support v_swap_b16.

### PR DESCRIPTION
**Pull request by broxigarchen**
support V_SWAP_B16 true16 encoding in asm/disasm for GFX11/12

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td [hunk 1]
```diff
@@ -729,7 +729,22 @@ def V_ACCVGPR_MOV_B32 : VOP1_Pseudo<"v_accvgpr_mov_b32", VOPProfileAccMov, [], 1
...
+def VOP_SWAP_I16 : VOPProfile_True16<VOP_I16_I16> {
+  let Outs32 = (outs VOPDstOperand_t16Lo128:$vdst,
+                     VOPSrcEncodedDstOperand_t16Lo128:$vdst1);
+  let Ins32 = (ins VOPSrcEncodedDstOperand_t16Lo128:$src0,
+                   VOPDstOperand_t16Lo128:$src1);
+  let Asm32 = " $vdst, $src0";
+}
+
 let SubtargetPredicate = isGFX11Plus in {
+  def V_SWAP_B16 : VOP1_Pseudo<"v_swap_b16", VOP_SWAP_I16, [], /* VOP1Only= */ 1> {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/VOP1Instructions.td:741]:
```suggestion
  def V_SWAP_B16 : VOP1_Pseudo<"v_swap_b16", VOP_SWAP_I16, [], /* VOP1Only= */true> {
```

**Review by arsenm — COMMENTED** [review]
Description is slightly misleading, this does not emit it from codegen and only handles the encoding/decoding

File: llvm/test/CodeGen/AMDGPU/v_swap_b16.ll [hunk 2]
```diff
@@ -0,0 +1,39 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -march=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX11 %s
```
Comment (Sisyph) [inline llvm/test/CodeGen/AMDGPU/v_swap_b16.ll:2]: It looks like v_mov_b16 and v_swap_b16 need to have the UseRealTrue16Insts predicate applied to them. We don't want to select them unless passing the feature flag mattr=+real-true16.

**Review by Sisyph — COMMENTED** [review]
Needs gfx12 assembler and disassembler tests.

Should resolve
#61622

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td (hunk 1 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td:741]: Hi Matt Thanks for comment. The other VOP1Only in this file is also using 1, so to just keep consistency I use 1 here as well.

File: llvm/test/CodeGen/AMDGPU/v_swap_b16.ll (hunk 2 above)
Comment (broxigarchen) [reply llvm/test/CodeGen/AMDGPU/v_swap_b16.ll]: done

Comment (broxigarchen) [reply llvm/test/CodeGen/AMDGPU/v_swap_b16.ll]: I will merge GFX12 v_swap_b16 mc change later in another patch

Comment (Sisyph) [reply llvm/test/CodeGen/AMDGPU/v_swap_b16.ll]: The gfx12 instructions were added in this patch, so the tests should also be. 

In addition to passing the mattr=+real-true16 in the test, we need a behavior change to emitted code and another test, of the default behavior. 
v_mov_b16 and v_swap_b16 need to have the UseRealTrue16Insts predicate applied to them in tablegen. We don't want to select them unless passing the feature flag mattr=+real-true16. 

We don't want to use any real true16 instruction by default. 
We should have 2 runlines for every 16 bit test. 1 with +real-true16, and one with no attribute specified, to check the default behavior (which is -real-true16).  In this case, it was clear in the prior version of the patch when the runline had no attribute specified the v_swap_b16 was still emitted.

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td:741]: Those should also be fixed, just use true here

**Comment by jayfoad** [comment]
Can you please split this into one patch which just adds MC (assembler/disassembler) support and tests, and then a second patch which adds CodeGen support and tests?

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td [hunk 3]
```diff
@@ -729,7 +729,20 @@ def V_ACCVGPR_MOV_B32 : VOP1_Pseudo<"v_accvgpr_mov_b32", VOPProfileAccMov, [], 1
...
```
Comment (DadSchoorse) [inline llvm/lib/Target/AMDGPU/VOP1Instructions.td:741]: Are you sure this opcode is VOP1 only? When I implemented support for `v_swap_b16` in our compiler, I tested it with VOP3 encoding and it worked fine. VOP3 v_swap_b16 is nessecary to address v128-v255.

Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: According to the docs this should be VOP1 only for GFX11/12. I am not sure if such usage is supported when VOP3 is not listed in supported format.  Hi @Sisyph @kosarev Do you know anything about this?

**Comment by broxigarchen** [comment]
> Can you please split this into one patch which just adds MC (assembler/disassembler) support and tests, and then a second patch which adds CodeGen support and tests?

Thanks for the comment! Just trim the patch to contain only MC changes

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td (hunk 1 above)
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: Done

File: llvm/test/CodeGen/AMDGPU/v_swap_b16.ll (hunk 2 above)
Comment (broxigarchen) [reply llvm/test/CodeGen/AMDGPU/v_swap_b16.ll]: Thanks for the comment! Just trim the patch to MC only and added GFX12 mc test

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td (hunk 3 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: It's VOP1 only in GFX11 and GFX12, and it does support v0-v255 for both operands.

Comment (DadSchoorse) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: That can't be true, VOP1 v_swap_b16  definitely operates on v0-v127.lo/.hi. There is no way to encode v128-v255 without VOP3, just like for other true16 instructions.

Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: Right, the lo/hi selector goes to the MSB of the operand fields, so can't encode registers beyond v127.

Comment (DadSchoorse) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: This is what I tested for what it's worth, and that instruction swaps v138.h and v139.h like expected:

https://gitlab.freedesktop.org/DadSchoorse/bvhre/-/blob/d1a4108b1547577860261d39a24ca4ea616c7664/ptest.cpp#L265

Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: From what I see, officially there's no VOP3 form, but I could ask ISA people to be sure. Which targets did you try this on, exactly?

Comment (DadSchoorse) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: I did all my gfx11 testing on navi31.

Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: OK, I did inquire on this and will update this thread as soon as I have a response. Thanks for reporting.

**Comment by Sisyph** [comment]
Can you please commit a separate patch first that contains the changes to llvm/test/MC/AMDGPU/gfx12_asm_vop1-fake16.s and llvm/test/MC/AMDGPU/gfx12_asm_vop1.s minus v_swap_b16? The files look fine, just adding that bulk test update should be separate from the functional change to v_swap_b16.

**Review by Sisyph — APPROVED** [review]
LGTM. If the VOP3 form is supported, we can add a follow up patch.

**Comment by broxigarchen** [comment]
> LGTM. If the VOP3 form is supported, we can add a follow up patch.

Thanks!

**Comment by broxigarchen** [comment]
Hi I think this PR is ready. Can anyone help to merge this PR? Thanks!

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td (hunk 3 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: The response from our HW guys reads that the VOP3 codes are not guaranteed to work correctly under all possible circumstances. Meaning, they should not be considered supported regardless of observed side effects.

Comment (DadSchoorse) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: Thanks for investigating, even if the answer is a bit disappointing.

**Pull request merged**